### PR TITLE
CI: Run benchmarks if PR is labeled with "run/benchmark"

### PR DESCRIPTION
**Pull request by seisman**
Support running benchmarks in PRs with the `run/benchmark` label.

Address comment https://github.com/GenericMappingTools/pygmt/pull/2952#discussion_r1443648845.

### CODE REVIEW

**Comment by seisman** [comment]
The slash command won't work until the PR is merged into main branch.

**Comment by weiji14** [comment]
> The slash command won't work until the PR is merged into main branch.

Actually, will CodSpeed be able to link the report to the PR branch if we trigger it via `repository_dispatch` instead of `pull_request`?

**Comment by seisman** [comment]
> > The slash command won't work until the PR is merged into main branch.
> 
> Actually, will CodSpeed be able to link the benchmark report to the PR branch if we trigger it via `repository_dispatch` instead of `pull_request`?

That's a good question. I guess the answer is no.

**Comment by weiji14** [comment]
> > Actually, will CodSpeed be able to link the benchmark report to the PR branch if we trigger it via `repository_dispatch` instead of `pull_request`?
> 
> That's a good question. I guess the answer is no.

How about we set up the workflow to run only when there's a `run-benchmark` label? Can try to set up an if-condition following https://stackoverflow.com/questions/62325286/run-github-actions-when-pull-requests-have-a-specific-label

**Comment by seisman** [comment]
> > > Actually, will CodSpeed be able to link the benchmark report to the PR branch if we trigger it via `repository_dispatch` instead of `pull_request`?
> > 
> > 
> > That's a good question. I guess the answer is no.
> 
> How about we set up the workflow to run only when there's a `run-benchmark` label? Can try to set up an if-condition following https://stackoverflow.com/questions/62325286/run-github-actions-when-pull-requests-have-a-specific-label

Sounds a clever solution.

**Comment by seisman** [comment]
The benchmark workflow is trigged by the "run-benchmark" label, see https://github.com/GenericMappingTools/pygmt/actions/runs/7431110786?pr=2958.

**Comment by seisman** [comment]
I feel triggering a workflow based on the PR label is much better than the slash commands. Some thoughts:

1. Migrate the other two slash commands `/format` and `/test-gmt-dev` to triggering by labels
2. Instead of using labels like `run-benchmark`, maybe labels `run/benchmark`, `run/format`, `run/test-gmt-dev` are better?

**Comment by seisman** [comment]
Hmm, the workflow is canceled because I added the 'maintenance' label.

**Comment by weiji14** [comment]
> 1. Migrate the other two slash commands `/format` and `/test-gmt-dev` to triggering by labels

Yeah, I was thinking of using it for `/test-gmt-dev`, but maybe not for `/format` since the labels can only be added by contributors/maintainers with triage permissions or above, so new contributors won't be able to use it. 

> 2. Instead of using labels like `run-benchmark`, maybe labels `run/benchmark`, `run/format`, `run/test-gmt-dev` are better?

Sounds good.

File: .github/workflows/benchmarks.yml
Comment (seisman) [inline .github/workflows/benchmarks.yml:31]: Actually I'm unsure if the workflow will run in the main branch, because it's a push event not a pull_request event.

Comment (seisman) [reply .github/workflows/benchmarks.yml:31]: Tried in my own fork. It doesn't work in the main branch.

Comment (seisman) [reply .github/workflows/benchmarks.yml]: Changed the `if` condition to:
```
if: github.repository == 'GenericMappingTools/pygmt' && (github.event_name != 'pull_request' || contains(github.event.pull_request.labels.*.name, 'run/benchmark'))
```
Now it works in my own fork, see the changes between my fork and the pygmt repository: https://github.com/GenericMappingTools/pygmt/compare/main...seisman:pygmt:main and the workflow was triggered when pushing in the main branch (https://github.com/seisman/pygmt/actions/workflows/benchmarks.yml).

**Comment by seisman** [comment]
It's unclear why I can't see the flame graphs.

File: .github/workflows/benchmarks.yml
Comment (weiji14) [inline .github/workflows/benchmarks.yml:15]: Should we document in `.github/PULL_REQUEST_TEMPLATE.md` or somewhere for contributors to request that the `run/benchmark` label gets added to PRs that might affect performance? Or just rely on the maintainers to set the label for now?

**Review by weiji14 — COMMENTED** [review]
> It's unclear why I can't see the flame graphs.

Hmm yeah, are you getting this `Execution profile not available` message too at https://codspeed.io/GenericMappingTools/pygmt/branches/slash-benchmark?

image

If I try another branch (e.g. https://codspeed.io/GenericMappingTools/pygmt/branches/alias/function), it looks ok, though sometimes the flame graphs can take some time to render.

File: .github/workflows/benchmarks.yml
Comment (seisman) [reply .github/workflows/benchmarks.yml:15]: I prefer to leave the jobs to the maintainers. We probably will document it in the contributors guides when addressing https://github.com/GenericMappingTools/pygmt/issues/1113.

Comment (weiji14) [reply .github/workflows/benchmarks.yml:15]: Ok, or we can put it in `doc/maintenance.md` somewhere. Not urgent though, since we aren't running the benchmarks that often.

**Review by weiji14 — APPROVED** [review]
Thanks @seisman! Have secretly wanted this label-based trigger mechanism in PyGMT for a while now to be honest :grinning:

**Comment by weiji14** [comment]
Quick question: Do we want to remove the `run/benchmark` label after a PR is merged, or keep the label on?

**Comment by seisman** [comment]
I'm OK with either. Maybe we should keep the label so that we know which branches have performance reports.

**Comment by seisman** [comment]
> Quick question: Do we want to remove the `run/benchmark` label after a PR is merged, or keep the label on?

Actually, we should remove the 'run/benchmark' label. We usually label a PR as "needs-review", then "final-review-call" and remove the "final-review-call" label before merging. All these actions will trigger the benchmarks workflow if the `run/benchmark` label is kept.

**Comment by weiji14** [comment]
> > Quick question: Do we want to remove the `run/benchmark` label after a PR is merged, or keep the label on?
> 
> Actually, we should remove the 'run/benchmark' label. We usually label a PR as "needs-review", then "final-review-call" and remove the "final-review-call" label before merging. All these actions will trigger the benchmarks workflow if the `run/benchmark` label is kept.

Owh, but we would need to remove all those labels at the same time right? Maybe better to set an if-condition so that the benchmark workflow is not re-triggered after a PR is already merged?

**Comment by seisman** [comment]
> Owh, but we would need to remove all those labels at the same time right?

We need to remove `run/benchmark` label as early as possible to avoid unnecessary runs.



> Maybe better to set an if-condition so that the benchmark workflow is not re-triggered after a PR is already merged?

It may make the `if`-condition too complicated, which is already very long.

**Comment by weiji14** [comment]
> > Owh, but we would need to remove all those labels at the same time right?
> 
> We need to remove `run/benchmark` label as early as possible to avoid unnecessary runs.

Ok, let's go with this. Remove the `run/benchmark` label before the other labels are removed. Same goes for the `run/test-gmt-dev` label.

**Pull request merged**
